1.19.3 compatibility

Spigot 1.19.3-3616
BigDoors 0.1.8.40

On server start-up
```
[08:21:04] [Server thread/INFO]: [BigDoors] Enabling BigDoors vAlpha 0.1.8.40
[08:21:04] [Server thread/INFO]: [BigDoors] Power Block Types:
[08:21:04] [Server thread/INFO]: [BigDoors]  - GOLD_BLOCK
[08:21:04] [Server thread/INFO]: [BigDoors] Blacklisted materials:
[08:21:04] [Server thread/INFO]: [BigDoors]  - BEDROCK
[08:21:04] [Server thread/INFO]: [BigDoors]  - END_PORTAL
[08:21:04] [Server thread/INFO]: [BigDoors]  - END_PORTAL_FRAME
[08:21:04] [Server thread/INFO]: [BigDoors]  - NETHER_PORTAL
[08:21:04] [Server thread/INFO]: [BigDoors] Whitelisted materials:
[08:21:04] [Server thread/INFO]: [BigDoors]  - REDSTONE_BLOCK
[08:21:04] [Server thread/INFO]: [BigDoors] DestroyListed materials:
[08:21:04] [Server thread/INFO]: [BigDoors]  - LAVA
[08:21:04] [Server thread/INFO]: [BigDoors]  - WATER
[08:21:04] [Server thread/INFO]: [BigDoors]  - SNOW
[08:21:04] [Server thread/INFO]: [BigDoors]  - FERN
[08:21:04] [Server thread/INFO]: [BigDoors]  - GRASS
[08:21:04] [Server thread/INFO]: [BigDoors]  - TALL_GRASS
[08:21:04] [Server thread/INFO]: [BigDoors]  - SEAGRASS
[08:21:04] [Server thread/INFO]: [BigDoors]  - TALL_SEAGRASS
[08:21:04] [Server thread/WARN]: [BigDoors] ╔═══════════════════════════════════════════════════════╗
[08:21:04] [Server thread/ERROR]: [BigDoors] java.lang.Exception: Failed to generate NMS code! Please contact pim16aap2!
        at nl.pim16aap2.bigDoors.codegeneration.FallbackGeneratorManager.generateFallingBlockFactory(FallbackGeneratorManager.java:75)
        at nl.pim16aap2.bigDoors.codegeneration.FallbackGeneratorManager.getFallingBlockFactory(FallbackGeneratorManager.java:40)
        at nl.pim16aap2.bigDoors.BigDoors.compatibleMCVer(BigDoors.java:816)
        at nl.pim16aap2.bigDoors.BigDoors.onEnable(BigDoors.java:187)
        at org.bukkit.plugin.java.JavaPlugin.setEnabled(JavaPlugin.java:264)
        at org.bukkit.plugin.java.JavaPluginLoader.enablePlugin(JavaPluginLoader.java:342)
        at org.bukkit.plugin.SimplePluginManager.enablePlugin(SimplePluginManager.java:480)
        at org.bukkit.craftbukkit.v1_19_R2.CraftServer.enablePlugin(CraftServer.java:539)
        at org.bukkit.craftbukkit.v1_19_R2.CraftServer.enablePlugins(CraftServer.java:453)
        at net.minecraft.server.MinecraftServer.loadWorld0(MinecraftServer.java:587)
        at net.minecraft.server.MinecraftServer.loadLevel(MinecraftServer.java:412)
        at net.minecraft.server.dedicated.DedicatedServer.e(DedicatedServer.java:250)
        at net.minecraft.server.MinecraftServer.w(MinecraftServer.java:973)
        at net.minecraft.server.MinecraftServer.lambda$0(MinecraftServer.java:301)
        at java.base/java.lang.Thread.run(Thread.java:1589)
Caused by: java.lang.ExceptionInInitializerError
        at nl.pim16aap2.bigDoors.codegeneration.FallbackGeneratorManager.getMappingsVersion(FallbackGeneratorManager.java:83)
        at nl.pim16aap2.bigDoors.codegeneration.FallbackGeneratorManager.generateFallingBlockFactory(FallbackGeneratorManager.java:55)
        ... 14 more
Caused by: java.lang.NullPointerException: Expected 1 method, but got 0 for input: [public static net.minecraft.world.level.block.state.IBlockData net.minecraft.nbt.GameProfileSerializer#[*](net.minecraft.nbt.NBTTagCompound)]. excluding superclasses, excluding interfaces
Found methods: []
        at nl.pim16aap2.bigDoors.reflection.ReflectionBackend.findMethod(ReflectionBackend.java:298)
        at nl.pim16aap2.bigDoors.reflection.MethodFinder$TypedMethodFinder.get(MethodFinder.java:218)
        at nl.pim16aap2.bigDoors.codegeneration.ReflectionRepository.<clinit>(ReflectionRepository.java:268)
        ... 16 more
```

--- a/changelog.txt
+++ b/changelog.txt
@@ -1,4 +1,5 @@
 Alpha 0.1.8.41
+- Add support for MC 1.19.3.
 - Add support for GriefDefender v2. Thanks, haroldrc, for the heads-up!
 - Add support for Medieval Factions 5. Thanks, OGVampire, for the suggestion!
 - Add config option to refund doors on delete. Thanks, LoneTiger, for the suggestion!

--- a/core/pom.xml
+++ b/core/pom.xml
@@ -152,7 +152,7 @@
         </dependency>
 
         <dependency>
-            <groupId>com.github.cryptomorin</groupId>
+            <groupId>com.github.PimvanderLoos</groupId>
             <artifactId>XSeries</artifactId>
             <version>${dependency.version.xseries}</version>
             <scope>compile</scope>
@@ -294,7 +294,7 @@
                     </relocations>
                     <filters>
                         <filter>
-                            <artifact>com.github.cryptomorin:XSeries</artifact>
+                            <artifact>com.github.PimvanderLoos:XSeries</artifact>
                             <includes>
                                 <include>com/cryptomorin/xseries/XMaterial*</include>
                             </includes>

--- a/core/pom.xml
+++ b/core/pom.xml
@@ -145,6 +145,13 @@
         </dependency>
 
         <dependency>
+            <groupId>nl.pim16aap2.BigDoors</groupId>
+            <artifactId>v1_19_R2</artifactId>
+            <version>0.1.8.41-ALPHA</version>
+            <scope>compile</scope>
+        </dependency>
+
+        <dependency>
             <groupId>com.github.cryptomorin</groupId>
             <artifactId>XSeries</artifactId>
             <version>${dependency.version.xseries}</version>

--- a/core/src/main/java/nl/pim16aap2/bigDoors/BigDoors.java
+++ b/core/src/main/java/nl/pim16aap2/bigDoors/BigDoors.java
@@ -17,6 +17,7 @@ import nl.pim16aap2.bigDoors.NMS.FallingBlockFactory_V1_18_R1;
 import nl.pim16aap2.bigDoors.NMS.FallingBlockFactory_V1_18_R2;
 import nl.pim16aap2.bigDoors.NMS.FallingBlockFactory_V1_19_R1;
 import nl.pim16aap2.bigDoors.NMS.FallingBlockFactory_V1_19_R1_1;
+import nl.pim16aap2.bigDoors.NMS.FallingBlockFactory_V1_19_R2;
 import nl.pim16aap2.bigDoors.codegeneration.FallbackGeneratorManager;
 import nl.pim16aap2.bigDoors.compatibility.FakePlayerCreator;
 import nl.pim16aap2.bigDoors.compatibility.ProtectionCompatManager;
@@ -811,6 +812,9 @@ public class BigDoors extends JavaPlugin implements Listener
                     fabf = new FallingBlockFactory_V1_19_R1();
                 else
                     fabf = new FallingBlockFactory_V1_19_R1_1();
+                break;
+            case "v1_19_R2":
+                fabf = new FallingBlockFactory_V1_19_R2();
                 break;
             default:
                 if (config.allowCodeGeneration())

--- a/core/src/main/java/nl/pim16aap2/bigDoors/codegeneration/EntityFallingBlockClassGenerator.java
+++ b/core/src/main/java/nl/pim16aap2/bigDoors/codegeneration/EntityFallingBlockClassGenerator.java
@@ -20,6 +20,7 @@ import java.lang.reflect.Field;
 import java.lang.reflect.InvocationTargetException;
 import java.lang.reflect.Method;
 import java.util.Collections;
+import java.util.Objects;
 
 import static net.bytebuddy.implementation.FixedValue.value;
 import static net.bytebuddy.implementation.MethodCall.construct;
@@ -111,7 +112,6 @@ final class EntityFallingBlockClassGenerator extends ClassGenerator
         builder = addAuxiliaryMethods(builder);
         builder = addTickMethod(builder);
         builder = addCrashReportMethod(builder);
-
         finishBuilder(builder);
     }
 
@@ -213,12 +213,25 @@ final class EntityFallingBlockClassGenerator extends ClassGenerator
                     baseObject.generated$loadTileEntityData(compound);
             }, ILoadDataDelegation.class));
 
+        final MethodCall deserializedInvocation;
+        if (methodIBlockDataDeserializer.getParameterCount() == 1)
+        {
+            deserializedInvocation = invoke(methodIBlockDataDeserializer)
+                .withMethodCall(invoke(methodNBTTagCompoundGetCompound).onArgument(0).with("BlockState"));
+        }
+        else if (methodIBlockDataDeserializer.getParameterCount() == 2)
+        {
+            Objects.requireNonNull(methodWorldReaderHolderLookup);
+            deserializedInvocation = invoke(methodIBlockDataDeserializer)
+                .withMethodCall(invoke(methodWorldReaderHolderLookup).onField(fieldNMSWorld).with(objectRegistryBlock))
+                .withMethodCall(invoke(methodNBTTagCompoundGetCompound).onArgument(0).with("BlockState"));
+        }
+        else
+            throw new IllegalStateException("Method IBlockDataDeserializer has unexpected number of parameters!");
+
         builder = builder
             .define(methodLoadData)
-            .intercept(invoke(methodIBlockDataDeserializer)
-                           .withMethodCall(invoke(methodNBTTagCompoundGetCompound)
-                                               .onArgument(0).with("BlockState"))
-                           .setsField(named(FIELD_BLOCK)).andThen(
+            .intercept(deserializedInvocation.setsField(named(FIELD_BLOCK)).andThen(
                     invoke(methodNBTTagCompoundGetInt)
                         .onArgument(0).with("Time").setsField(fieldTicksLived)).andThen(
                     invoke(named(METHOD_NAME_LOAD_DATA_CONDITIONAL))

--- a/core/src/main/java/nl/pim16aap2/bigDoors/codegeneration/ReflectionASMAnalyzers.java
+++ b/core/src/main/java/nl/pim16aap2/bigDoors/codegeneration/ReflectionASMAnalyzers.java
@@ -10,9 +10,7 @@ import java.lang.reflect.Modifier;
 import java.util.List;
 import java.util.Objects;
 
-import static nl.pim16aap2.bigDoors.reflection.ReflectionBuilder.findConstructor;
-import static nl.pim16aap2.bigDoors.reflection.ReflectionBuilder.findField;
-import static nl.pim16aap2.bigDoors.reflection.ReflectionBuilder.findMethod;
+import static nl.pim16aap2.bigDoors.reflection.ReflectionBuilder.*;
 
 @SuppressWarnings("SameParameterValue") //
 final class ReflectionASMAnalyzers

--- a/core/src/main/java/nl/pim16aap2/bigDoors/codegeneration/ReflectionRepository.java
+++ b/core/src/main/java/nl/pim16aap2/bigDoors/codegeneration/ReflectionRepository.java
@@ -2,6 +2,7 @@ package nl.pim16aap2.bigDoors.codegeneration;
 
 import com.cryptomorin.xseries.XMaterial;
 import nl.pim16aap2.bigDoors.BigDoors;
+import nl.pim16aap2.bigDoors.reflection.ParameterGroup;
 import nl.pim16aap2.bigDoors.reflection.ReflectionBuilder;
 import org.bukkit.Bukkit;
 import org.bukkit.Location;
@@ -12,11 +13,13 @@ import org.bukkit.block.data.BlockData;
 import org.bukkit.event.entity.CreatureSpawnEvent;
 import org.bukkit.material.MaterialData;
 
+import javax.annotation.Nullable;
 import java.lang.reflect.Array;
 import java.lang.reflect.Constructor;
 import java.lang.reflect.Field;
 import java.lang.reflect.Method;
 import java.lang.reflect.Modifier;
+import java.util.Arrays;
 import java.util.Collections;
 import java.util.List;
 
@@ -52,6 +55,11 @@ final class ReflectionRepository
     public static final Class<?> classCraftBlockData;
     public static final Class<?> classBlockStateEnum;
     public static final Class<?> classEntityTypes;
+    public static final @Nullable Class<?> classHolderGetter;
+    public static final @Nullable Class<?> classRegistries;
+    public static final Class<?> classResourceKey;
+    public static final Class<?> classIWorldReader; // net.minecraft.world.level.LevelReader
+    public static final @Nullable Class<?> classHolderLookup;
 
     public static final Class<?> classEnumBlockState;
     public static final Class<?> classEnumMoveType;
@@ -129,6 +137,8 @@ final class ReflectionRepository
     public static final Method methodEnumOrdinal;
     public static final Method methodArrayGetIdx;
     public static final Method methodGetClass;
+    public static final @Nullable Method methodWorldReaderHolderLookup;
+    private static final @Nullable Method methodRegistriesRegisterNew;
 
     public static final Field fieldTileEntityData;
     public static final Field fieldTicksLived;
@@ -137,6 +147,8 @@ final class ReflectionRepository
     public static final Field fieldEntityTypeFallingBlock;
 
     public static final List<Field> fieldsVec3D;
+
+    public static final @Nullable Object objectRegistryBlock;
 
     static
     {
@@ -188,7 +200,11 @@ final class ReflectionRepository
         classEnumBlockState = findClass(nmsBase + "BlockStateEnum",
                                         "net.minecraft.world.level.block.state.properties.BlockStateEnum").get();
         classEntityTypes = findClass("net.minecraft.world.entity.EntityTypes").get();
-
+        classHolderGetter = findClass("net.minecraft.core.HolderGetter").setNullable().get();
+        classRegistries = findClass("net.minecraft.core.registries.Registries").setNullable().get();
+        classResourceKey = findClass("net.minecraft.resources.ResourceKey").get();
+        classIWorldReader = findClass("net.minecraft.world.level.IWorldReader").get();
+        classHolderLookup = findClass("net.minecraft.core.HolderLookup").setNullable().get();
 
         cTorPrivateNMSFallingBlockEntity =
             findConstructor().inClass(classEntityFallingBlock)
@@ -262,18 +278,27 @@ final class ReflectionRepository
                                                  .withReturnType(classNBTTagCompound)
                                                  .withModifiers(Modifier.PUBLIC, Modifier.STATIC)
                                                  .withParameters(classIBlockData).get();
-        methodIBlockDataDeserializer = findMethod().inClass(classGameProfileSerializer)
-                                                   .withReturnType(classIBlockData)
-                                                   .withModifiers(Modifier.PUBLIC, Modifier.STATIC)
-                                                   .withParameters(classNBTTagCompound).get();
+        methodIBlockDataDeserializer =
+            findMethod().inClass(classGameProfileSerializer)
+                        .withReturnType(classIBlockData)
+                        .withModifiers(Modifier.PUBLIC, Modifier.STATIC)
+                        .withParameters(ParameterGroup.builder()
+                                                      .withOptionalParameters(classHolderGetter == null ?
+                                                                              new Class<?>[0] :
+                                                                              new Class<?>[]{classHolderGetter})
+                                                      .withRequiredParameters(classNBTTagCompound).construct()).get();
         methodCraftBockDataFromNMSBlockData = findMethod().inClass(classCraftBlockData).withName("fromData")
                                                           .withParameters(classIBlockData).get();
         methodBlockBaseGetItem = findMethod().inClass(classBlockBase).withReturnType(classNMSItem)
                                              .withoutParameters()
                                              .withModifiers(Modifier.ABSTRACT, Modifier.PUBLIC).get();
+        // In 1.19.3, there are 2 methods that fit this lookup. They both do basically the same thing
+        // and for our use-case, it doesn't matter which one we use.
         methodSetIBlockDataHolderState = ReflectionBuilder.findMethod().inClass(classIBlockDataHolder)
+                                                          .findMultiple().atLeast(1).atMost(2)
                                                           .withReturnType(Object.class)
-                                                          .withParameters(classIBlockState, Comparable.class).get();
+                                                          .withParameters(classIBlockState, Comparable.class)
+                                                          .get().get(0);
         methodCraftMagicNumbersGetMaterial = findMethod()
             .inClass(classCraftMagicNumbers).withName("getMaterial").withParameters(classIBlockData).get();
         methodGetItemType = findMethod().inClass(MaterialData.class).withName("getItemType").get();
@@ -333,10 +358,24 @@ final class ReflectionRepository
         methodLocY = locationMethods[1];
         methodLocZ = locationMethods[2];
 
+        if (classRegistries == null)
+        {
+            methodRegistriesRegisterNew = null;
+            objectRegistryBlock = null;
+            methodWorldReaderHolderLookup = null;
+        }
+        else
+        {
+            methodRegistriesRegisterNew = findMethod().inClass(classRegistries).withReturnType(classResourceKey)
+                                                      .withParameters(String.class).get();
+            methodRegistriesRegisterNew.setAccessible(true);
+            objectRegistryBlock = invoke(methodRegistriesRegisterNew, null, "block");
+            methodWorldReaderHolderLookup = findMethod().inClass(classIWorldReader).withReturnType(classHolderLookup)
+                                                        .withParameters(classResourceKey).setNullable().get();
+        }
+
         fieldEntityTypeFallingBlock = ReflectionASMAnalyzers.getEntityTypeFallingBlock(classEntityTypes,
                                                                                        cTorPrivateNMSFallingBlockEntity);
-
-
         fieldTileEntityData = findField().inClass(classEntityFallingBlock).ofType(classNBTTagCompound)
                                          .withModifiers(Modifier.PUBLIC).get();
         fieldTicksLived = findField().inClass(classEntityFallingBlock).ofType(int.class)
@@ -359,5 +398,21 @@ final class ReflectionRepository
     public static Class<?> asArrayType(Class<?> clz)
     {
         return Array.newInstance(clz, 0).getClass();
+    }
+
+    @SuppressWarnings("SameParameterValue")
+    private static <T> T invoke(Method method, @Nullable Object source, Object... params)
+    {
+        method.setAccessible(true);
+        try
+        {
+            //noinspection unchecked
+            return (T) method.invoke(source, params);
+        }
+        catch (Exception e)
+        {
+            throw new RuntimeException("Could not invoke method '" + method + "' on source: '" + source +
+                                           "' with params: " + Arrays.toString(params), e);
+        }
     }
 }

--- a/core/src/main/java/nl/pim16aap2/bigDoors/reflection/ParameterGroup.java
+++ b/core/src/main/java/nl/pim16aap2/bigDoors/reflection/ParameterGroup.java
@@ -136,6 +136,14 @@ public class ParameterGroup
     }
 
     /**
+     * Creates a new builder for this class.
+     */
+    public static Builder builder()
+    {
+        return new Builder();
+    }
+
+    /**
      * Represents a builder for {@link ParameterGroup}s.
      */
     public static class Builder

--- a/core/src/main/java/nl/pim16aap2/bigDoors/reflection/asm/ASMDebugging.java
+++ b/core/src/main/java/nl/pim16aap2/bigDoors/reflection/asm/ASMDebugging.java
@@ -1,18 +1,46 @@
 package nl.pim16aap2.bigDoors.reflection.asm;
 
 import nl.pim16aap2.bigDoors.util.Constants;
+import org.jetbrains.annotations.NotNull;
+import org.objectweb.asm.AnnotationVisitor;
+import org.objectweb.asm.ClassReader;
+import org.objectweb.asm.ClassVisitor;
+import org.objectweb.asm.ClassWriter;
+import org.objectweb.asm.FieldVisitor;
 import org.objectweb.asm.MethodVisitor;
 import org.objectweb.asm.util.Textifier;
+import org.objectweb.asm.util.TraceAnnotationVisitor;
+import org.objectweb.asm.util.TraceFieldVisitor;
 import org.objectweb.asm.util.TraceMethodVisitor;
 
 import java.io.IOException;
 import java.io.PrintWriter;
 import java.lang.reflect.Executable;
 
+@SuppressWarnings("unused")
 public final class ASMDebugging
 {
     private ASMDebugging()
     {
+    }
+
+    public static void printClassBytecode(Class<?> clz)
+    {
+        try
+        {
+            final PrintWriter pw = new PrintWriter(System.out);
+            pw.println("Going to print bytecode for class: '" + clz + "'\n");
+            final int debugTypes = DebugType.ANNOTATION | DebugType.METHOD | DebugType.FIELD;
+
+            final ClassReader cr = ASMUtil.getClassReader(clz);
+            final ClassWriter cw = new ClassWriter(cr, ClassWriter.COMPUTE_MAXS);
+            final DebugClassVisitor cv = new DebugClassVisitor(cw, debugTypes, pw);
+            cr.accept(cv, 0);
+        }
+        catch (Exception e)
+        {
+            throw new RuntimeException("Failed to print bytecode for class '" + clz + "'", e);
+        }
     }
 
     public static void printMethodBytecode(Executable executable)
@@ -34,9 +62,9 @@ public final class ASMDebugging
         ASMUtil.processMethod(executable, mv -> getDebugMethodVisitor(mv, printWriter), null);
     }
 
-    public static MethodVisitor getDebugMethodVisitor(MethodVisitor methodVisitor, PrintWriter printWriter)
+    private static MethodVisitor getDebugMethodVisitor(MethodVisitor visitor, PrintWriter printWriter)
     {
-        return new TraceMethodVisitor(methodVisitor, new Textifier(Constants.ASM_API_VER)
+        return new TraceMethodVisitor(visitor, new Textifier(Constants.ASM_API_VER)
         {
             @Override
             public void visitMethodEnd()
@@ -47,8 +75,76 @@ public final class ASMDebugging
         });
     }
 
-    public static MethodVisitor getDebugMethodVisitor(MethodVisitor methodVisitor)
+    private static FieldVisitor getDebugFieldVisitor(FieldVisitor visitor, PrintWriter printWriter)
     {
-        return getDebugMethodVisitor(methodVisitor, new PrintWriter(System.out));
+        return new TraceFieldVisitor(visitor, new Textifier(Constants.ASM_API_VER)
+        {
+            @Override
+            public void visitFieldEnd()
+            {
+                print(printWriter);
+                printWriter.flush();
+            }
+        });
+    }
+
+    private static AnnotationVisitor getDebugAnnotationVisitor(AnnotationVisitor visitor, PrintWriter printWriter)
+    {
+        return new TraceAnnotationVisitor(visitor, new Textifier(Constants.ASM_API_VER)
+        {
+            @Override
+            public void visitAnnotationEnd()
+            {
+                print(printWriter);
+                printWriter.flush();
+            }
+        });
+    }
+
+    private static final class DebugClassVisitor extends ClassVisitor
+    {
+        private final int debugTypes;
+        private final PrintWriter printWriter;
+
+        DebugClassVisitor(@NotNull ClassWriter cw, int debugTypes, PrintWriter printWriter)
+        {
+            super(Constants.ASM_API_VER, cw);
+            this.debugTypes = debugTypes;
+            this.printWriter = printWriter;
+        }
+
+        @Override
+        public AnnotationVisitor visitAnnotation(String desc, boolean visible)
+        {
+            final AnnotationVisitor av = super.visitAnnotation(desc, visible);
+            if ((debugTypes & DebugType.ANNOTATION) > 0)
+                return getDebugAnnotationVisitor(av, printWriter);
+            return av;
+        }
+
+        @Override
+        public FieldVisitor visitField(int access, String name, String desc, String signature, Object value)
+        {
+            final FieldVisitor fv = super.visitField(access, name, desc, signature, value);
+            if ((debugTypes & DebugType.FIELD) > 0)
+                return getDebugFieldVisitor(fv, printWriter);
+            return fv;
+        }
+
+        @Override
+        public MethodVisitor visitMethod(int access, String name, String desc, String signature, String[] exceptions)
+        {
+            final MethodVisitor mv = super.visitMethod(access, name, desc, signature, exceptions);
+            if ((debugTypes & DebugType.METHOD) > 0)
+                return getDebugMethodVisitor(mv, printWriter);
+            return mv;
+        }
+    }
+
+    public static final class DebugType
+    {
+        public static final int METHOD          = 1;
+        public static final int FIELD           = 1 << 1;
+        public static final int ANNOTATION      = 1 << 2;
     }
 }

--- a/core/src/main/java/nl/pim16aap2/bigDoors/reflection/asm/ASMUtil.java
+++ b/core/src/main/java/nl/pim16aap2/bigDoors/reflection/asm/ASMUtil.java
@@ -183,7 +183,7 @@ public final class ASMUtil
         return cv.replacementVisitor;
     }
 
-    private static ClassReader getClassReader(Class<?> clz)
+    public static ClassReader getClassReader(Class<?> clz)
         throws IOException
     {
         final String classAsPath = getClassName(clz) + ".class";

--- a/core/src/main/java/nl/pim16aap2/bigDoors/util/Util.java
+++ b/core/src/main/java/nl/pim16aap2/bigDoors/util/Util.java
@@ -506,8 +506,8 @@ public final class Util
         }
         catch (Exception e)
         {
-            BigDoors.get().getMyLogger().warn("Could not determine material of mat: " + mat.name());
-            e.printStackTrace();
+            BigDoors.get().getMyLogger().warn("Could not determine material of mat: " + mat.name() +
+                                                  ". Reason: " + e.getMessage());
             return null;
         }
     }

--- a/hooks/hook-griefdefender-1/pom.xml
+++ b/hooks/hook-griefdefender-1/pom.xml
@@ -21,7 +21,7 @@
         </dependency>
 
         <dependency>
-            <groupId>com.github.cryptomorin</groupId>
+            <groupId>com.github.PimvanderLoos</groupId>
             <artifactId>XSeries</artifactId>
             <scope>provided</scope>
         </dependency>

--- a/hooks/hook-griefdefender-2/pom.xml
+++ b/hooks/hook-griefdefender-2/pom.xml
@@ -28,7 +28,7 @@
         </dependency>
 
         <dependency>
-            <groupId>com.github.cryptomorin</groupId>
+            <groupId>com.github.PimvanderLoos</groupId>
             <artifactId>XSeries</artifactId>
             <scope>provided</scope>
         </dependency>

--- a/hooks/hook-konquest/pom.xml
+++ b/hooks/hook-konquest/pom.xml
@@ -21,7 +21,7 @@
         </dependency>
 
         <dependency>
-            <groupId>com.github.cryptomorin</groupId>
+            <groupId>com.github.PimvanderLoos</groupId>
             <artifactId>XSeries</artifactId>
             <scope>provided</scope>
         </dependency>

--- a/nms/pom.xml
+++ b/nms/pom.xml
@@ -34,7 +34,7 @@
 
     <dependencies>
         <dependency>
-            <groupId>com.github.cryptomorin</groupId>
+            <groupId>com.github.PimvanderLoos</groupId>
             <artifactId>XSeries</artifactId>
             <scope>provided</scope>
         </dependency>

--- a/nms/pom.xml
+++ b/nms/pom.xml
@@ -29,6 +29,7 @@
         <module>v1_18_R2</module>
         <module>v1_19_R1</module>
         <module>v1_19_R1_1</module>
+        <module>v1_19_R2</module>
     </modules>
 
     <dependencies>

--- a/nms/v1_19_R2/pom.xml
+++ b/nms/v1_19_R2/pom.xml
@@ -1,0 +1,36 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0"
+         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+    <packaging>jar</packaging>
+    <artifactId>v1_19_R2</artifactId>
+
+    <parent>
+        <groupId>nl.pim16aap2.BigDoors</groupId>
+        <artifactId>nms</artifactId>
+        <version>0.1.8.41-ALPHA</version>
+    </parent>
+
+    <properties>
+        <maven.compiler.source>8</maven.compiler.source>
+        <maven.compiler.target>8</maven.compiler.target>
+        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+    </properties>
+
+    <dependencies>
+        <dependency>
+            <groupId>org.spigotmc</groupId>
+            <artifactId>spigot</artifactId>
+            <version>1.19.3-R0.1-SNAPSHOT</version>
+            <scope>provided</scope>
+        </dependency>
+
+        <dependency>
+            <groupId>nl.pim16aap2.BigDoors</groupId>
+            <artifactId>nms-api</artifactId>
+            <version>0.1.8.41-ALPHA</version>
+            <scope>provided</scope>
+        </dependency>
+    </dependencies>
+</project>

--- a/nms/v1_19_R2/src/main/java/nl/pim16aap2/bigDoors/NMS/CustomCraftFallingBlock_V1_19_R2.java
+++ b/nms/v1_19_R2/src/main/java/nl/pim16aap2/bigDoors/NMS/CustomCraftFallingBlock_V1_19_R2.java
@@ -1,0 +1,99 @@
+package nl.pim16aap2.bigDoors.NMS;
+
+import org.bukkit.Material;
+import org.bukkit.Server;
+import org.bukkit.block.data.BlockData;
+import org.bukkit.craftbukkit.v1_19_R2.block.data.CraftBlockData;
+import org.bukkit.craftbukkit.v1_19_R2.entity.CraftEntity;
+import org.bukkit.craftbukkit.v1_19_R2.util.CraftMagicNumbers;
+import org.bukkit.entity.EntityType;
+import org.bukkit.entity.FallingBlock;
+import org.bukkit.util.EulerAngle;
+import org.bukkit.util.Vector;
+import org.jetbrains.annotations.NotNull;
+
+public class CustomCraftFallingBlock_V1_19_R2 extends CraftEntity implements FallingBlock, CustomCraftFallingBlock
+{
+    CustomCraftFallingBlock_V1_19_R2(final Server server, final CustomEntityFallingBlock_V1_19_R2 entity)
+    {
+        super((org.bukkit.craftbukkit.v1_19_R2.CraftServer) server, entity);
+        setVelocity(new Vector(0, 0, 0));
+        setDropItem(false);
+    }
+
+    @Override
+    public CustomEntityFallingBlock_V1_19_R2 getHandle()
+    {
+        return (CustomEntityFallingBlock_V1_19_R2) entity;
+    }
+
+    @Override
+    public boolean isOnGround()
+    {
+        return false;
+    }
+
+    @Override
+    public String toString()
+    {
+        return "CraftFallingBlock";
+    }
+
+    @Override
+    public @NotNull EntityType getType()
+    {
+        return EntityType.FALLING_BLOCK;
+    }
+
+    @Override
+    @Deprecated
+    public @NotNull Material getMaterial()
+    {
+        return CraftMagicNumbers.getMaterial(getHandle().i()).getItemType();
+    }
+
+    @Override
+    public @NotNull BlockData getBlockData()
+    {
+        return CraftBlockData.fromData(this.getHandle().i());
+    }
+
+    @Override
+    public boolean getDropItem()
+    {
+        return false;
+    }
+
+    @Override
+    public void setDropItem(final boolean drop)
+    {
+    }
+
+    @Override
+    public boolean canHurtEntities()
+    {
+        return false;
+    }
+
+    @Override
+    public void setHurtEntities(final boolean hurtEntities)
+    {
+    }
+
+    @Override
+    public void setTicksLived(final int value)
+    {
+        super.setTicksLived(value);
+        getHandle().b = value;
+    }
+
+    @Override
+    public void setHeadPose(EulerAngle pose)
+    {
+    }
+
+    @Override
+    public void setBodyPose(EulerAngle eulerAngle)
+    {
+    }
+}

--- a/nms/v1_19_R2/src/main/java/nl/pim16aap2/bigDoors/NMS/CustomEntityFallingBlock_V1_19_R2.java
+++ b/nms/v1_19_R2/src/main/java/nl/pim16aap2/bigDoors/NMS/CustomEntityFallingBlock_V1_19_R2.java
@@ -1,0 +1,105 @@
+package nl.pim16aap2.bigDoors.NMS;
+
+import net.minecraft.CrashReportSystemDetails;
+import net.minecraft.core.BlockPosition;
+import net.minecraft.core.registries.Registries;
+import net.minecraft.nbt.GameProfileSerializer;
+import net.minecraft.nbt.NBTTagCompound;
+import net.minecraft.server.level.WorldServer;
+import net.minecraft.world.damagesource.DamageSource;
+import net.minecraft.world.entity.EntityTypes;
+import net.minecraft.world.entity.EnumMoveType;
+import net.minecraft.world.entity.item.EntityFallingBlock;
+import net.minecraft.world.level.block.state.IBlockData;
+import net.minecraft.world.phys.Vec3D;
+import org.bukkit.craftbukkit.v1_19_R2.CraftWorld;
+import org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason;
+
+/**
+ * v1_19_R1 implementation of {@link CustomEntityFallingBlock}.
+ *
+ * @author Pim
+ * @see CustomEntityFallingBlock
+ */
+public class CustomEntityFallingBlock_V1_19_R2 extends EntityFallingBlock implements CustomEntityFallingBlock
+{
+    private IBlockData block;
+
+    public CustomEntityFallingBlock_V1_19_R2(final org.bukkit.World world, final double d0, final double d1,
+                                             final double d2, final IBlockData iblockdata)
+    {
+        super(EntityTypes.F, ((CraftWorld) world).getHandle());
+        block = iblockdata;
+
+        this.f(d0, d1, d2);
+        super.b = 0;
+        super.aq = false;
+        super.Q = true;
+        this.e(true);
+        this.f(new Vec3D(0.0D, 0.0D, 0.0D));
+        this.a(new BlockPosition(this.dk(), this.dm(), this.dq()));
+        spawn();
+    }
+
+    public void spawn()
+    {
+        ((WorldServer)super.s).addWithUUID(this, SpawnReason.CUSTOM);
+    }
+
+    @Override
+    public void l()
+    {
+        if (block.h())
+            ak();
+        else
+        {
+            a(EnumMoveType.a, di());
+            if (++b > 12000)
+                ak();
+
+            f(di().d(0.9800000190734863D, 1.0D, 0.9800000190734863D));
+        }
+    }
+
+    @Override
+    public boolean a(float f, float f1, DamageSource damagesource)
+    {
+        return false;
+    }
+
+    @Override
+    protected void b(final NBTTagCompound nbttagcompound)
+    {
+        nbttagcompound.a("BlockState", GameProfileSerializer.a(block));
+        nbttagcompound.a("Time", b);
+        nbttagcompound.a("DropItem", false);
+        nbttagcompound.a("HurtEntities", aq);
+        nbttagcompound.a("FallHurtAmount", 0.0f);
+        nbttagcompound.a("FallHurtMax", 0);
+        if (d != null)
+            nbttagcompound.a("TileEntityData", d);
+    }
+
+    @Override
+    protected void a(final NBTTagCompound nbttagcompound)
+    {
+        block = GameProfileSerializer.a(super.s.a(Registries.e), nbttagcompound.p("BlockState"));
+        b = nbttagcompound.h("Time");
+
+        if (nbttagcompound.b("TileEntityData", 10))
+            super.d = nbttagcompound.p("TileEntityData");
+    }
+
+    @Override
+    public void a(final CrashReportSystemDetails crashreportsystemdetails)
+    {
+        super.a(crashreportsystemdetails);
+        crashreportsystemdetails.a("Animated BigDoors block with state: ", block.toString());
+    }
+
+    @Override
+    public IBlockData i()
+    {
+        return block;
+    }
+}

--- a/nms/v1_19_R2/src/main/java/nl/pim16aap2/bigDoors/NMS/FallingBlockFactory_V1_19_R2.java
+++ b/nms/v1_19_R2/src/main/java/nl/pim16aap2/bigDoors/NMS/FallingBlockFactory_V1_19_R2.java
@@ -1,0 +1,34 @@
+package nl.pim16aap2.bigDoors.NMS;
+
+import net.minecraft.core.BlockPosition;
+import net.minecraft.world.level.block.state.BlockBase;
+import net.minecraft.world.level.block.state.BlockBase.Info;
+import net.minecraft.world.level.block.state.IBlockData;
+import org.bukkit.Bukkit;
+import org.bukkit.Location;
+import org.bukkit.Material;
+import org.bukkit.World;
+import org.bukkit.craftbukkit.v1_19_R2.CraftWorld;
+
+public class FallingBlockFactory_V1_19_R2 implements FallingBlockFactory
+{
+    @Override
+    public CustomCraftFallingBlock fallingBlockFactory(Location loc, NMSBlock block, byte matData, Material mat)
+    {
+        IBlockData blockData = ((NMSBlock_V1_19_R2) block).getMyBlockData();
+        CustomEntityFallingBlock_V1_19_R2 fBlockNMS
+            = new CustomEntityFallingBlock_V1_19_R2(loc.getWorld(), loc.getX(), loc.getY(), loc.getZ(), blockData);
+        CustomCraftFallingBlock_V1_19_R2 entity = new CustomCraftFallingBlock_V1_19_R2(Bukkit.getServer(), fBlockNMS);
+        entity.setCustomName("BigDoorsEntity");
+        entity.setCustomNameVisible(false);
+        return entity;
+    }
+
+    @Override
+    public NMSBlock nmsBlockFactory(World world, int x, int y, int z)
+    {
+        final Info blockInfo =
+            Info.a((BlockBase) ((CraftWorld) world).getHandle().a_(new BlockPosition(x, y, z)).b());
+        return new NMSBlock_V1_19_R2(world, x, y, z, blockInfo);
+    }
+}

--- a/nms/v1_19_R2/src/main/java/nl/pim16aap2/bigDoors/NMS/NMSBlock_V1_19_R2.java
+++ b/nms/v1_19_R2/src/main/java/nl/pim16aap2/bigDoors/NMS/NMSBlock_V1_19_R2.java
@@ -1,0 +1,214 @@
+package nl.pim16aap2.bigDoors.NMS;
+
+import com.cryptomorin.xseries.XMaterial;
+import net.minecraft.core.BlockPosition;
+import net.minecraft.core.EnumDirection;
+import net.minecraft.world.item.Item;
+import net.minecraft.world.level.block.Block;
+import net.minecraft.world.level.block.BlockRotatable;
+import net.minecraft.world.level.block.EnumBlockRotation;
+import net.minecraft.world.level.block.state.BlockBase;
+import net.minecraft.world.level.block.state.IBlockData;
+import nl.pim16aap2.bigDoors.util.DoorDirection;
+import nl.pim16aap2.bigDoors.util.NMSUtil;
+import nl.pim16aap2.bigDoors.util.RotateDirection;
+import org.bukkit.Location;
+import org.bukkit.Material;
+import org.bukkit.World;
+import org.bukkit.block.BlockFace;
+import org.bukkit.block.data.MultipleFacing;
+import org.bukkit.block.data.Waterlogged;
+import org.bukkit.block.data.type.Fence;
+import org.bukkit.craftbukkit.v1_19_R2.CraftWorld;
+import org.bukkit.craftbukkit.v1_19_R2.block.data.CraftBlockData;
+
+import java.util.Objects;
+import java.util.Set;
+
+public class NMSBlock_V1_19_R2 extends BlockBase implements NMSBlock
+{
+    private IBlockData blockData;
+    private CraftBlockData craftBlockData;
+    private final XMaterial xmat;
+    private Location loc;
+
+    public NMSBlock_V1_19_R2(World world, int x, int y, int z, Info blockInfo)
+    {
+        super(blockInfo);
+
+        loc = new Location(world, x, y, z);
+
+        craftBlockData = (CraftBlockData) world.getBlockAt(x, y, z).getBlockData();
+        if (craftBlockData instanceof Waterlogged)
+            ((Waterlogged) craftBlockData).setWaterlogged(false);
+
+        updateBlockData();
+
+        xmat = XMaterial.matchXMaterial(world.getBlockAt(x, y, z).getType().toString()).orElse(XMaterial.BEDROCK);
+    }
+
+    @Override
+    public boolean canRotate()
+    {
+        return craftBlockData instanceof MultipleFacing;
+    }
+
+    @Override
+    public void rotateBlock(RotateDirection rotDir)
+    {
+        EnumBlockRotation rot;
+        switch (rotDir)
+        {
+        case CLOCKWISE:
+            rot = EnumBlockRotation.b;
+            break;
+        case COUNTERCLOCKWISE:
+            rot = EnumBlockRotation.d;
+            break;
+        default:
+            rot = EnumBlockRotation.a;
+        }
+        blockData = blockData.a(rot);
+        updateCraftBlockData();
+    }
+
+    private void updateBlockData()
+    {
+        blockData = craftBlockData.getState();
+    }
+
+    private void updateCraftBlockData()
+    {
+        this.craftBlockData = CraftBlockData.fromData(blockData);
+    }
+
+    @Override
+    public void putBlock(Location loc)
+    {
+        this.loc = loc;
+
+        if (craftBlockData instanceof MultipleFacing)
+            updateCraftBlockDataMultipleFacing();
+
+        ((CraftWorld) Objects.requireNonNull(loc.getWorld()))
+            .getHandle().a(new BlockPosition(loc.getBlockX(), loc.getBlockY(), loc.getBlockZ()), blockData, 1);
+    }
+
+    private void updateCraftBlockDataMultipleFacing()
+    {
+        final Set<BlockFace> allowedFaces = ((MultipleFacing) craftBlockData).getAllowedFaces();
+        allowedFaces.forEach(
+            (blockFace) ->
+            {
+                final org.bukkit.block.Block otherBlock =
+                    loc.clone().add(blockFace.getModX(), blockFace.getModY(), blockFace.getModZ()).getBlock();
+                final org.bukkit.block.data.BlockData otherData = otherBlock.getBlockData();
+
+                if (blockFace.equals(BlockFace.UP))
+                    ((MultipleFacing) craftBlockData).setFace(blockFace, true);
+                else if (otherBlock.getType().isSolid())
+                {
+
+                    ((MultipleFacing) craftBlockData).setFace(blockFace, true);
+
+                    final boolean isOtherMultipleFacing = otherData instanceof MultipleFacing;
+                    final boolean materialMatch = otherBlock.getType().equals(xmat.parseMaterial());
+                    final boolean areBothFence = craftBlockData instanceof Fence && otherData instanceof Fence;
+
+                    if (isOtherMultipleFacing && (materialMatch || areBothFence))
+                    {
+                        final Set<BlockFace> otherAllowedFaces = ((MultipleFacing) otherData).getAllowedFaces();
+                        if (otherAllowedFaces.contains(blockFace.getOppositeFace()))
+                        {
+                            ((MultipleFacing) otherData).setFace(blockFace.getOppositeFace(), true);
+                            otherBlock.setBlockData(otherData);
+                        }
+                    }
+                }
+                else
+                    ((MultipleFacing) craftBlockData).setFace(blockFace, false);
+            });
+        this.updateBlockData();
+    }
+
+    @Override
+    public void rotateVerticallyInDirection(DoorDirection openDirection)
+    {
+        NMSUtil.rotateVerticallyInDirection(openDirection, craftBlockData);
+        this.updateBlockData();
+    }
+
+    @Override
+    public void rotateBlockUpDown(boolean ns)
+    {
+        EnumDirection.EnumAxis axis = blockData.c(BlockRotatable.g);
+        EnumDirection.EnumAxis newAxis = axis;
+        switch (axis)
+        {
+        case a:
+            newAxis = ns ? EnumDirection.EnumAxis.a : EnumDirection.EnumAxis.b;
+            break;
+        case b:
+            newAxis = ns ? EnumDirection.EnumAxis.c : EnumDirection.EnumAxis.a;
+            break;
+        case c:
+            newAxis = ns ? EnumDirection.EnumAxis.b : EnumDirection.EnumAxis.c;
+            break;
+        }
+        blockData = blockData.a(BlockRotatable.g, newAxis);
+        this.updateCraftBlockData();
+    }
+
+    @Override
+    public void rotateCylindrical(RotateDirection rotDir)
+    {
+        if (rotDir.equals(RotateDirection.CLOCKWISE))
+            blockData = blockData.a(EnumBlockRotation.b);
+        else
+            blockData = blockData.a(EnumBlockRotation.d);
+        this.updateCraftBlockData();
+    }
+
+    /**
+     * Gets the IBlockData (NMS) of this block.
+     *
+     * @return The IBlockData (NMS) of this block.
+     */
+    IBlockData getMyBlockData()
+    {
+        return blockData;
+    }
+
+    @Override
+    public String toString()
+    {
+        return blockData.toString();
+    }
+
+    @Override
+    public void deleteOriginalBlock(boolean applyPhysics)
+    {
+        final World world = Objects.requireNonNull(loc.getWorld());
+        if (!applyPhysics)
+        {
+            world.getBlockAt(loc).setType(Material.AIR, false);
+        }
+        else
+        {
+            world.getBlockAt(loc).setType(Material.CAVE_AIR, false);
+            world.getBlockAt(loc).setType(Material.AIR, true);
+        }
+    }
+
+    @Override
+    public Item l()
+    {
+        return null;
+    }
+
+    @Override
+    protected Block p()
+    {
+        return null;
+    }
+}

--- a/pom.xml
+++ b/pom.xml
@@ -26,7 +26,7 @@
         <dependency.version.asm>9.3</dependency.version.asm>
         <dependency.version.mockito>4.5.1</dependency.version.mockito>
         <dependency.version.junit>5.8.2</dependency.version.junit>
-        <dependency.version.xseries>9.1.0</dependency.version.xseries>
+        <dependency.version.xseries>bf8bc2c34b</dependency.version.xseries>
     </properties>
 
     <repositories>
@@ -45,7 +45,7 @@
     <dependencyManagement>
         <dependencies>
             <dependency>
-                <groupId>com.github.cryptomorin</groupId>
+                <groupId>com.github.PimvanderLoos</groupId>
                 <artifactId>XSeries</artifactId>
                 <version>${dependency.version.xseries}</version>
                 <scope>provided</scope>

--- a/pom.xml
+++ b/pom.xml
@@ -26,7 +26,7 @@
         <dependency.version.asm>9.3</dependency.version.asm>
         <dependency.version.mockito>4.5.1</dependency.version.mockito>
         <dependency.version.junit>5.8.2</dependency.version.junit>
-        <dependency.version.xseries>8.8.0</dependency.version.xseries>
+        <dependency.version.xseries>9.1.0</dependency.version.xseries>
     </properties>
 
     <repositories>

